### DISCUSSION TITLE
How can I bind `jj` to `escape` in insert mode.

### DISCUSSION THREAD

**Discussion by anhsirk0** [meow-edit/meow#168] [Q&A]
title says it.
How can I bind `jj` to `escape` in insert mode.

**Comment by DogLooksGood** [comment]
Hi, there's no such feature.

I'm not sure this will help or not.
https://github.com/emacsorphanage/key-chord

**Reply by ml729** [reply]
Unfortunately, key-chord mode's chords do not work when defining kmacros. Here is an alternative way that supports kmacros, based off of [this](https://emacs.stackexchange.com/a/20024) stackexchange answer:
``` elisp
(setq meow-two-char-escape-sequence "jk")
(setq meow-two-char-escape-delay 0.5)

(defun meow--two-char-exit-insert-state (s)
  (when (meow-insert-mode-p)
    (let ((modified (buffer-modified-p)))
      (insert (elt s 0))
      (let* ((second-char (elt s 1))
             (event
              (if defining-kbd-macro
                  (read-event nil nil)
              (read-event nil nil meow-two-char-escape-delay))))
        (when event
          (if (and (characterp event) (= event second-char))
              (progn
                (backward-delete-char 1)
                (set-buffer-modified-p modified)
                (meow--execute-kbd-macro "<escape>"))
            (push event unread-command-events)))))))

(defun meow-two-char-exit-insert-state ()
  (interactive)
  (meow--two-char-exit-insert-state meow-two-char-escape-sequence))

(define-key meow-insert-state-keymap (substring meow-two-char-escape-sequence 0 1)
  #'meow-two-char-exit-insert-state)
```
(Minor warning: If you're recording a kmacro and press the first letter of `meow-two-char-escape-sequence`, and then immediately after press a key that calls `kmacro-end-or-call-macro` or `kmacro-end-and-call-macro` etc., this will cause a kmacro that infinitely calls itself. Fortunately, Meow's kmacro commands prevent kmacros from ending in insert state, so if you stick to using those you don't have to worry about this.)

**Reply by anhsirk0** [reply]
Thanks , Its a nice solution , but I am currently using `capslock` as `escape`. 
PS: I am using meow continuously for months and its great, I like how numbers repeat actions very intuitive. Thanks for your hard work on this project :)

**Reply by grszkthfr** [reply]
How hard would it be to implement something like [`evil-escape`](https://github.com/syl20bnr/evil-escape)?

**Reply by ml729** [reply]
It doesn’t seem that bad. If you or anyone else is interested in implementing a Meow version, it looks like the bulk of the work is done by the functions `evil-escape-pre-command-hook` and `evil-escape-p`. Then there are per state functions like `evil-escape--escape-normal-state`, and `evil-escape--escape-motion-state` that call the right quit function per mode/context. 

But personally I'm ok with using `C-g` and `meow-quit` for the situations handled by `evil-escape`.

**Comment by anhsirk0** [comment]
Is there any other way to exit insert mode without pressing `escape` like with `C-q`?
`escape` key is too far for me.

**Reply by DogLooksGood** [reply]
yes, bind `meow-insert-exit` to whatever you like in `meow-insert-state-keymap`. It's just a normal keymap, `define-key` works.

https://github.com/meow-edit/meow/blob/master/meow-keymap.el#L61

**Reply by anhsirk0** [reply]
Thanks!!
